### PR TITLE
Fix work with types in `pattern_merge_ranges + __hetero_tag<_BackendTag>`

### PR DESCRIPTION
**Pull request by SergeyKopienko**
In this PR we fix not quite correct work with `range_difference_t` types of source and result ranges in `pattern_merge_ranges(__hetero_tag<_BackendTag>`.
Also we add new test case for `data_in_in_out_lim` mode - we test the case when output range is empty.
And we fix one error in the tests inside `struct merge_checker_fn::operator()` which crashes in this new test case.

### CODE REVIEW

File: test/parallel_api/ranges/std_ranges_merge.pass.cpp [hunk 1]
```diff
@@ -24,41 +24,31 @@ struct merge_checker_fn
     auto operator()(_R1&& r_1, _R2&& r_2, _ROut&& r_out, Comp comp = {}, Proj1 proj1 = {}, Proj2 proj2 = {})
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply test/parallel_api/ranges/std_ranges_merge.pass.cpp:24]: For now we works only with `random-access` ranges.
